File: 2019/08-Aug/08-18-type-equality/php/equality.php
<?php

/**
 * Demonstrate how equality works in PHP
 * @author Andrew Jarombek
 * @since 8/17/2019
 */

require_once "Yarn.php";

// Similar to JavaScript, PHP has type coercion.  PHP is in general considered a loosely typed language.
assert(2 == '2');
assert("2" == 2);
assert(2 == 2);

assert("adding love 2 heart database tomorrow" != "when I finish this article");

assert(true == "1");
assert(false == "0");

// Just like JavaScript, this can cause strange behavior
assert("00" == "0000");
assert(null == array());

// However, you could say it's not "as" loosely typed as JavaScript.  In JavaScript, [10] == 10 returns true.
assert([10] != 10);

// Just like JavaScript, in PHP == allows type coercion when comparing values while  === disallows type coercion.
assert(2 === 2);

// This causes type coercion to fail.  A string and an integer aren't held in the same memory location.
assert(2 !== "2");

// PHP does have a concept of references, however I can't find anything on reference equality for primitive types.
// PHP does not have the concept of pointers.
$name = "Andy";
$nameAgain =& $name;

// Since both $name and $nameAgain reference the same string, altering one impacts both.
$name = $name . " Jarombek";

assert($name == $nameAgain);
assert($name == "Andy Jarombek");
assert($nameAgain == "Andy Jarombek");

// When testing equality of objects, == tests for value equality and === tests for reference equality.  These operators
// work without any explicit definitions and without operator overloading.
$yarn1 = new Yarn("Polyester", "Pitter Patter", 210);
$yarn2 = $yarn1;
$yarn3 = new Yarn("Polyester", "Pitter Patter", 210);
$yarn4 = new Yarn("Polyester", "Vanilla", 70);

assert($yarn1 == $yarn2);
assert($yarn1 === $yarn2);

assert($yarn2 == $yarn3);
assert($yarn2 !== $yarn3);

assert($yarn3 != $yarn4);
assert($yarn3 !== $yarn4);

File: 2019/08-Aug/08-18-type-equality/cpp/equality.cpp
/**
 * Demonstrate how equality works in C++.
 * @author Andrew Jarombek
 * @date 8/17/2019
 */

#include <iostream>
#include <typeinfo>
#include <cassert>
#include "WrappingPaper.h"

using namespace std;

/**
 * Struct representing a ball of yarn.
 */
typedef struct Yarn {
    string fiber;
    string color;
    int yards;

    /**
     * Overload the == operator to test value equality between two Yarn structs.
     * @param other Another Yarn struct to test equality with.
     * @return true if the two Yarn structs have equal values, false otherwise.
     */
    bool operator == (Yarn& other) {
        return fiber == other.fiber
             && color == other.color
             && yards == other.yards;
    }

    /**
     * Overload the != operator to test value inequality between two Yarn structs.
     * @param other Another Yarn struct to test equality with.
     * @return true if the two Yarn structs DON'T have equal values, false otherwise.
     */
    bool operator != (Yarn& other) {
        return !(*this == other);
    }

} Yarn;

int main() {
    // Basic type equality is mostly the same in C++ compared to C.  However, the == operator returns a boolean value
    // instead of an integer.
    int two = 2;
    int twoAgain = 2;

    // 'auto' can be replaced with 'bool'
    auto twosAreEqual = two == twoAgain;

    assert(two == twoAgain);

    cout << typeid(twosAreEqual).name() << endl; // 'b' for boolean
    assert(twosAreEqual);

    // In C++, comparing a string to a character array correctly determines value equivalence.
    // This wasn't the case in C.  This is because the == operator is overloaded for strings in C++.
    string name = "Andy";
    string nameAgain = "Andy";
    string lastName = "Jarombek";
    char nameArray[] = "Andy";

    assert(name == nameAgain);
    assert(name == nameArray);
    assert(name != lastName);

    // You can also use compare() to test for string equality, although it isn't recommended.
    assert(name.compare(nameAgain) == 0);

    // Similar to C, there is no == operator for structs by default.  Unlike C, in C++ the == operator can be
    // overloaded to work with two structs of the same type.
    Yarn yarn1 = {"Polyester", "Pitter Patter", 210};
    Yarn& yarn2 = yarn1;
    Yarn yarn3 = {"Polyester", "Pitter Patter", 210};
    Yarn yarn4 = {"Polyester", "Vanilla", 70};

    // Use == and != to test the structs for value equality.
    assert(yarn1 == yarn2);
    assert(yarn2 == yarn3);
    assert(yarn3 != yarn4);

    // Use == and != on the pointers of the structs to test for reference equality.
    assert(&yarn1 == &yarn2);
    assert(&yarn2 != &yarn3);
    assert(&yarn3 != &yarn4);

    // The last assertion can be made at compile time instead of runtime.
    static_assert(
        &yarn3 != &yarn4,
        "Assertion Error: Different valued structs have the same memory location at compile time."
    );

    // Classes can also overload the == and != operators.
    string brand = "Hallmark";
    string pattern = "Disney Princess";
    string unknown = "Unknown";

    WrappingPaper wrappingPaper1 = {&brand, &pattern};
    WrappingPaper* wrappingPaper2 = &wrappingPaper1;
    WrappingPaper wrappingPaper3 = {&brand, &pattern};
    WrappingPaper wrappingPaper4 = {&unknown, &unknown};

    // Use == and != to test the objects for value equality.
    assert(wrappingPaper1 == *wrappingPaper2);
    assert(*wrappingPaper2 == wrappingPaper3);
    assert(wrappingPaper3 != wrappingPaper4);

    // Use == and != on the pointers of the objects to test for reference equality.
    assert(&wrappingPaper1 == wrappingPaper2);
    assert(wrappingPaper2 != &wrappingPaper3);
    assert(&wrappingPaper3 != &wrappingPaper4);

    // The last assertion can be made at compile time instead of runtime.
    static_assert(
        &wrappingPaper3 != &wrappingPaper4,
        "Assertion Error: Different valued objects have the same memory location at compile time."
    );
}